src/types/mapping.rs

```diff
@@ -105,10 +105,34 @@ impl PyMapping {
 }
 
 impl<'v> PyTryFrom<'v> for PyMapping {
+    // for Python < 3.10 or if using the Py_LIMITED_API, call into python to check
+    // isinstance(value, collections.abc.Mapping) to determine downcastability
+    #[cfg(any(Py_LIMITED_API, Py_3_6, Py_3_7, Py_3_8, Py_3_9))]
+    fn try_from<V: Into<&'v PyAny>>(value: V) -> Result<&'v PyMapping, PyDowncastError<'v>> {
+        let value = value.into();
+        let is_mapping = Python::with_gil(|py| {
+            let builtins = PyModule::import(py, "builtins")?;
+            let mapping_abc = PyModule::import(py, "collections.abc")?.getattr("Mapping")?;
+            builtins
+                .getattr("isinstance")?
+                .call1((value, mapping_abc))?
+                .extract::<bool>()
+        });
+        if is_mapping.unwrap_or(false) {
+            unsafe { Ok(<PyMapping as PyTryFrom>::try_from_unchecked(value)) }
+        } else {
+            Err(PyDowncastError::new(value, "Mapping"))
+        }
+    }
+
+    // for Python >= 3.10 and not using the Py_LIMITED_API, check Py_TPFLAGS_MAPPING to determine
+    // downcastability
+    #[cfg(not(any(Py_LIMITED_API, Py_3_6, Py_3_7, Py_3_8, Py_3_9)))]
     fn try_from<V: Into<&'v PyAny>>(value: V) -> Result<&'v PyMapping, PyDowncastError<'v>> {
         let value = value.into();
         unsafe {
-            if ffi::PyMapping_Check(value.as_ptr()) != 0 {
+            let ty = ffi::Py_TYPE(value);
+            if ffi::PyType_HasFeature(ty, fft::Py_TPFLAGS_MAPPING) != 0 {
```

Comment:
As a sanity check, I wonder if there is any evidence we can collect (or tests we can write) that this check is functionally equivalent to the slow path? If there's any lingering concern that they may be different, we probably need to always use the slow path.

Reply:
Good call - I'll give this some thought. I still think this is the right way going forward, but in the mean time (until 3.9 EOL I guess...) the behavior should be the same.

Reply:
I think it's not just 3.9 EOL - we also need to be sure that we won't have behaviour differences between abi3 and non-abi3 builds (just perf improvements).